srml/support/src/traits.rs

```diff
@@ -0,0 +1,197 @@
+	/// deleted: specifically `FreeBalance`. Furthermore, `OnFreeBalanceZero` callback
+	/// is invoked, giving a chance to external modules to cleanup data associated with
+	/// the deleted account.
+	///
+	/// `system::AccountNonce` is also deleted if `ReservedBalance` is also zero (it also gets
+	/// collapsed to zero if it ever becomes less than `ExistentialDeposit`.
+	fn free_balance(who: &AccountId) -> Self::Balance;
+
+	/// The amount of the balance of a given account that is externally reserved; this can still get
+	/// slashed, but gets slashed last of all.
+	///
+	/// This balance is a 'reserve' balance that other subsystems use in order to set aside tokens
+	/// that are still 'owned' by the account holder, but which are suspendable. (This is different
+	/// and wholly unrelated to the `Bondage` system used in the staking module.)
+	///
+	/// When this balance falls below the value of `ExistentialDeposit`, then this 'reserve account'
+	/// is deleted: specifically, `ReservedBalance`.
+	///
+	/// `system::AccountNonce` is also deleted if `FreeBalance` is also zero (it also gets
+	/// collapsed to zero if it ever becomes less than `ExistentialDeposit`.
+	fn reserved_balance(who: &AccountId) -> Self::Balance;
+
+	// PUBLIC MUTABLES (DANGEROUS)
+
+	/// Deducts up to `value` from the combined balance of `who`, preferring to deduct from the
+	/// free balance. This function cannot fail.
+	///
+	/// As much funds up to `value` will be deducted as possible. If this is less than `value`,
+	/// then `Some(remaining)` will be returned. Full completion is given by `None`.
+	fn slash(who: &AccountId, value: Self::Balance) -> Option<Self::Balance>;
+
+	/// Adds up to `value` to the free balance of `who`.
+	///
+	/// If `who` doesn't exist, nothing is done and an Err returned.
+	fn reward(who: &AccountId, value: Self::Balance) -> result::Result<(), &'static str>;
+
+	/// Adds up to `value` to the free balance of `who`.
+	///
+	/// If `who` doesn't exist, it is created
+	///
+	/// Returns if the account was successfully updated or update has led to killing of the account.
+	fn reward_creating(who: &AccountId, value: Self::Balance) -> UpdateBalanceOutcome;
+
+	/// Adds up to `value` to the free balance of `who`.
+	///
+	/// If `who` doesn't exist, nothing is done and an Err returned.
+	///
+	/// NOTE: This assumes that the total stake remains unchanged after this operation.
+	fn reward_with_no_stake_increase(who: &AccountId, value: Self::Balance) -> result::Result<(), &'static str>;
+
+	/// Adds up to `value` to the free balance of `who`.
+	///
+	/// If `who` doesn't exist, it is created
+	///
+	/// Returns if the account was successfully updated or update has led to killing of the account.
+	///
+	/// NOTE: This assumes that the total stake remains unchanged after this operation.
+	fn reward_with_no_stake_increase_creating(who: &AccountId, value: Self::Balance) -> UpdateBalanceOutcome;
```

Comment:
This appears to be a synonym for `increase_free_balance_creating`, which has a very important caveat in its documentation missing here:

```
	/// Adds up to `value` to the free balance of `who`. If `who` doesn't exist, it is created.
	///
	/// This is a sensitive function since it circumvents any fees associated with account
	/// setup. Ensure it is only called by trusted code.
	///
	/// NOTE: This assumes that the total stake remains unchanged after this operation. If
	/// you mean to actually mint value into existence, then use `reward` instead.
	pub fn increase_free_balance_creating(who: &T::AccountId, value: T::Balance) -> UpdateBalanceOutcome {
		Self::set_free_balance_creating(who, Self::free_balance(who) + value)
	}
```

Reply:
I have used "reward" to mean a safe operation that doesn't need thinking about too deeply: no fees are skipped (the destination account doesn't exist, then the operation fails without introducing a potential crypto-economic flaw) and the total issued amount is increased.

I would prefer to keep it having that meaning, with the lower-level "increase_free_balance" functions having the accordingly less safe implications.